scala/ScalaWorker.scala

```diff
@@ -0,0 +1,220 @@
+
+  private def getFreePort(): Int = {
+    val sock = new ServerSocket(0)
+    val port = sock.getLocalPort
+    sock.close()
+    port
+  }
+
+  private var zincClient: ZincClient = _
+
+  private var zincPort: Int = 0
+
+  private var nailgunProcess: Process = _
+
+  private def attachShutdownHook() {
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      override def run() {
+        if (nailgunProcess != null) {
+          nailgunProcess.destroy()
+        }
+      }
+    })
+  }
+
+  private val serverOutput = new StringBuilder()
+
+  private def startServer(classpath: String): Unit = {
+    attachShutdownHook()
+    zincPort = getFreePort()
+
+    val logger = new ProcessLogger {
+      def buffer[T](fn: => T): T = fn
+      def err(s: => String): Unit = serverOutput.append(s).append("\n")
+      def out(s: => String): Unit = serverOutput.append(s).append("\n")
+    }
+
+    // Options copied from Nailgun.scala in Zinc
+    val options = List("-cp", classpath, "-server", "-Xms1024m", "-Xmx3g", "-XX:MaxPermSize=384m",
+      "-XX:ReservedCodeCacheSize=192m")
+    val cmd = "java" :: options ++ Seq(classOf[Nailgun].getName, s"$zincPort")
+    val builder = Process(cmd)
+    this.nailgunProcess = builder.run(logger)
+
+    serverArgs = cmd.mkString(" ")
+    zincClient = new ZincClient(port = zincPort)
+  }
+
+  private def awaitServer() {
+    var count = 0
+    while (!zincClient.serverAvailable && (count < 50)) {
+      try { Thread.sleep(100) } catch { case _: InterruptedException => }
+      count += 1
+    }
+  }
+
+  def main(args: Array[String]): Unit = {
+    if (args.contains("--persistent_worker")) {
+      startServer(args(0))
+      runPersistentWorker(args)
+    } else {
+      // This is a single invocation of the example that exits after it processed the request.
+      ZincMain.run(args, cwd = None)
+    }
+  }
+
+  private def listFiles(f: File): Seq[String] = {
+    val current = f.listFiles
+    val files = current.filter(_.isFile).map(_.getAbsolutePath)
+    val directories = current.filter(_.isDirectory)
+    files ++ directories.flatMap(listFiles)
+  }
+
+  // Extract a src jar to a temporary directory and return the list of extracted files
+  private def expandSrcJar(path: String): Seq[String] = {
+    val tempDir = Files.createTempDirectory(null).toFile
+    Seq("unzip", "-q", path, "-d", tempDir.getAbsolutePath).!!
+    listFiles(tempDir)
+  }
+
+  @throws[IOException]
+  private def runPersistentWorker(args: Array[String]) {
+    val originalStdOut = System.out
+    val originalStdErr = System.err
+
+    while (true) {
+      try {
+        val request = WorkRequest.parseDelimitedFrom(System.in)
+        if (request == null) {
+          return
+        }
+
+        inputs.clear()
+
+        for (input <- request.getInputsList().asScala) {
+          inputs.put(input.getPath(), input.getDigest().toStringUtf8())
+        }
+
+        val baos = new ByteArrayOutputStream()
+        var exitCode = 0
```

Comment:
can we have `val exitCode = try { ...` below rather than a var? var adds cognitive load to reviewing, which IMO should only be used when necessary for perf. In this case `val` is just as good.